### PR TITLE
Integrate gitlab auth

### PR DESCRIPTION
**Pull request by r-scheele**
[Feature]
`/gitlab/login` route requires cliend_id, and client_secret from gitlab - Users Settings > Applications
the route returns an authorization code, which is redirected to a callback url in our application. verified, and then return an access token to the client.
sample response:
```json
{
  "access_token": "f50e2bf44ff901c2f619cef768f306f2561431413f0362a55fba42ec94fd1c12",
  "expires_in": 7200
}
```

### CODE REVIEW

**Review by Youngestdev — CHANGES_REQUESTED** [review]
Great job, Habeeb. Left you a couple of comments.

File: app/config/config.py [hunk 1]
```diff
@@ -9,6 +9,8 @@ class Settings(BaseSettings):
     TEST_DATABASE_PATH: Optional[str] = ""
     OPAL_SERVER_DATA_URL: Optional[str] = ""
     ENVIRONMENT: Optional[str] = "production"
+    GITLAB_CLIENT_ID: Optional[str] = ""
+    GITLAB_CLIENT_SECRET: Optional[str] = ""
```
[the remark this answers is not in the document]
Comment (r-scheele) [reply app/config/config.py:13]: Yes, It was like that for testing. I'll comment the part we don't need out, and push again

File: app/server/auth/authenticate.py [hunk 2]
```diff
@@ -1,26 +1,57 @@
...
+            "client_secret": APP_SECRET,
+            "code": code,
+            "grant_type": "authorization_code",
+            "redirect_uri": "http://localhost:8080/api/auth/callback/gitlab",
+        },
+    )
+    json_res = res.json()
+
+    access_token, expires_in = json_res["access_token"], json_res["expires_in"]
+    return {"access_token": access_token, "expires_in": expires_in}
```
[the remark this answers is not in the document]
Comment (r-scheele) [reply app/server/auth/authenticate.py:49]: Yes yes, like the one from github. I'll modify

[the remark this answers is not in the document]
Comment (Youngestdev) [reply app/server/auth/authenticate.py]: Remove it instead. No need keeping unused code.

File: README.md [hunk 3]
```diff
@@ -56,12 +58,15 @@ Create a postgres database, called datasource <br />
...
   ```
 
-Run the application entry point:
+Run the application entry point - development mode:
+
+
 
 ```console
-$ python3 main.py
+$ uvicorn app.server.api:app --port 8080 --reload
```
Comment (Youngestdev) [inline README.md:66]: Replace it with:

```console
$ gunicorn -w 4 -k uvicorn.workers.UvicornWorker app.server.api:app --bind 0.0.0.0:8080
```

File: app/schemas/rules.py [hunk 4]
```diff
@@ -19,7 +19,8 @@ class Config:
             "example": {
                 "name": "Example",
                 "owner": "r-scheele",
-                "github_repo_url": "https://github.com/r-scheele/opal-policy-example",
+                "repo_url": "https://github.com/r-scheele/opal-policy-example",
+                # "github_repo_url": "https://gitlab.com/r-scheele/opal-policy-example",
```
Comment (Youngestdev) [inline app/schemas/rules.py:23]: remove redundant comment.

File: app/schemas/rules.py [hunk 5]
```diff
@@ -130,7 +131,8 @@ class Config:
...
```
Comment (Youngestdev) [inline app/schemas/rules.py:135]: remove comment if it's redundant.

**Review by Youngestdev — COMMENTED** [review]
What does the new change do?

File: README.md (hunk 3 above)
Comment (r-scheele) [reply README.md:66]: This doesn't reload the app, we need it for development don't we?

File: app/schemas/rules.py (hunk 4 above)
Comment (r-scheele) [reply app/schemas/rules.py:23]: Alright sir, sure

File: app/schemas/rules.py (hunk 5 above)
Comment (r-scheele) [reply app/schemas/rules.py:135]: sure

File: README.md (hunk 3 above)
Comment (Youngestdev) [reply README.md:66]: Oh ok. That's true.

**Review by Youngestdev — APPROVED** [review]

**Pull request merged**